File: generated_code/python/solution_98_2.py
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        """
        Determines if a binary tree is a valid Binary Search Tree (BST).
        
        Algorithm: Recursive Depth-First Search (DFS) with a valid range.
        - For each node, its value must be strictly greater than a `low` bound 
          and strictly less than a `high` bound.
        - When moving left, the upper bound updates to the current node's value.
        - When moving right, the lower bound updates to the current node's value.
        
        Complexity:
        - Time Complexity: O(N) where N is the number of nodes, as we visit each node exactly once.
        - Space Complexity: O(H) where H is the height of the tree, representing the recursion stack. 
                            In the worst case (skewed tree), H = O(N). In the best case, H = O(log N).
        """
        def validate(node: Optional[TreeNode], low: float, high: float) -> bool:
            # An empty tree/node is a valid BST
            if not node:
                return True
            
            # The current node's value must stay strictly within the allowed range
            if not (low < node.val < high):
                return False
            
            # Recursively validate the left and right subtrees with updated bounds
            return validate(node.left, low, node.val) and validate(node.right, node.val, high)
        
        # Initialize the boundary limits with negative and positive infinity to handle potential integer boundaries
        return validate(root, float('-inf'), float('inf'))

File: generated_code/cpp/solution_98_2.cpp
#include <optional>

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    bool isValidBST(TreeNode* root) {
        // Use std::optional<int> to handle the full range of int values safely without overflow risks.
        return validate(root, std::nullopt, std::nullopt);
    }

private:
    // Helper function that uses recursion with strict upper and lower bounds.
    // Time Complexity: O(N) where N is the number of nodes, as we visit each node exactly once.
    // Space Complexity: O(H) where H is the height of the tree, representing the recursion stack depth.
    bool validate(TreeNode* node, std::optional<int> low, std::optional<int> high) {
        // An empty tree/node is a valid BST.
        if (!node) {
            return true;
        }

        // The current node's value must be strictly greater than the lower bound (if it exists).
        if (low && node->val <= *low) {
            return false;
        }
        // The current node's value must be strictly less than the upper bound (if it exists).
        if (high && node->val >= *high) {
            return false;
        }

        // Recursively validate the left and right subtrees with updated bounds.
        // Left subtree values must be strictly less than the current node's value.
        // Right subtree values must be strictly greater than the current node's value.
        return validate(node->left, low, node->val) && validate(node->right, node->val, high);
    }
};